### PR TITLE
update checkpoints, chainwork, assumevalid for mainnet and testnet

### PR DESCRIPTION
**Pull request by patricklodder**
Opening this earlier than usual, so that we can keep an eye on this, because hash rate has significantly dropped the past few months [according to bitinfocharts](https://bitinfocharts.com/comparison/dogecoin-hashrate.html#3m) and I have confirmed this to be consistent with what my own monitoring indicates:

3-mo hashrate estimate from bitinfocharts

This means that we have to be a little bit more careful when setting a checkpoint. We can easily, if the proposed checkpoint from this PR does not turn out to be optimal, go back further in time and use an earlier block. However, it's always good to have an as-recent-as-possible checkpoint, so let's try this one first.

### mainnet

`ed7d266dcbd8bb8af80f9ccb8deb3e18f9cc3f6972912680feeb37b090f8cee0` is the new mainnet checkpoint at height `4303965`, with a total tx count of `86433645`.

Calculated the tx volume between block 3793538 and 4288126 which is exactly 1 year of volume to be 0.29 tx/s. It's recently been going up a bit again, but the long-term trend is definitely downward rather than upward.

Cumulative chainwork is `0x07dc8ab65fc32f953c4a`

### testnet

`af23c3e750bb4f2ce091235f006e7e4e2af453d4c866282e7870471dcfeb4382` is the new testnet checkpoint at height `3976284`, with a total tx count of `5353803`.

The tx volume did not significantly change (0.024 to 0.026) so that remains unchanged.

Cumulative chainwork is `0x3e3c33bc605e5d`

### Note to reviewers

Things that need to be checked:

1. You can check the correctness of my proposal by doing `dogecoin-cli getblock <blockhash>`, which will contain everything except the cumulative transaction count. The easiest way to get the tx count is by searching your debug.log for the `UpdateTip` containing the mentioned hash, which will list the cumulative transaction count under `tx=<number>`.
2. Make sure that the checkpoint is on the active chain. Normally I'd wait for 1440 confirmations, but I propose to wait a little longer this time in the light of the drop in hash power, and aim for 3x that - 4320 confirmations.
3. We have to make sure that there are no big chaintip forks between this checkpoint and merge of the PR, you can do this by inspecting `dogecoin-cli getchaintips`. Look for alternative chaintips longer than 1-2 blocks, as the most common ones are simply occurring when there is a timing issue between miners and happen a couple of times per day usually.

If any of these conditions are not satisfied after 4320 confirmations, I can prepare a new proposal.

### CODE REVIEW

**Comment by victorsk2019** [comment]
Is the objective of this action to facilitate increase in difficulty (as I understand it, increasing difficulty causes increase in hash rate).  Not long ago, there was > 500 Th/s hash rate when trading volume was high.  There seems to be some kind of correlation between trading volume and hash rate, or it could just be a coincidence 🤷‍♂️

**Comment by patricklodder** [comment]
The objective of this PR is to create a checkpoint that will help 1.14.6 nodes (when we release) to have a recent checkpoint (on top of the one we already provide) to compare the data they get to when they sync in case someone would try to trick them into accepting the wrong chain.

It does not increase the hash rate. The consideration is that if 20% of the hash rate suddenly went away, we wonder where it went. So we have to be careful when we hard code a checkpoint.

**Comment by victorsk2019** [comment]
I see, thank you for clarifying, good to know what it’s for :)

**Comment by victorsk2019** [comment]
> We aim to do this every release. Recent ones for: [1.14.3](https://github.com/dogecoin/dogecoin/pull/1731), [1.14.4](https://github.com/dogecoin/dogecoin/pull/2452), [1.14.5](https://github.com/dogecoin/dogecoin/pull/2662)

One question though: isn't the ultimate goal to keep hash rate as high as possible so as to reduce the risk of someone with greater hashing power to overtake the chain, as was the situation with BCH-BSV   dispute several years ago?

**Comment by patricklodder** [comment]
> isn't the ultimate goal to keep hash rate as high as possible?

Correct, but you reversed cause and effect earlier. It is not that higher difficulty increases hash rate, the difficulty is what changes, [per the algorithm](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/dogecoin.cpp#L41), when  the time between blocks changes, which is, when it happens structurally and under normal circumstances, often caused by a change in hardware running hashes.

Note that because we only know the hashes that were found and the time it took to find them, not the hashes that weren't found, we can only infer the hash rate with some reasonable (or unreasonable) precision by aggregating lots of data and working the statistics. Big moves, like the ~20% drop depicted above, are relatively easy to spot. The volatility in the chart does give you an idea about how unreliable the metric is though, and the algorithm we use actually makes it more volatile.

Bottom line, mining equipment that does scrypt hashing and actually submits the hashes to the blockchain decides the difficulty and from that we calculate the hash rate. It is in the end driven by the operation of hardware and unless I missed the birth of Skynet, machines don't build or operate themselves yet.

ezgif-1-2bbe75866b

**Comment by chromatic** [comment]
I'm reviewing this and so far so good. We need about 300 more confirmations though.

File: src/chainparams.cpp [hunk 1]
```diff
@@ -263,10 +264,10 @@ class CTestNetParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout = 0; // Disabled
 
         // The best chain should have at least this much work.
-        consensus.nMinimumChainWork = uint256S("0x00000000000000000000000000000000000000000000000000001030d1382ade");
+        consensus.nMinimumChainWork = uint256S("000000000000000000000000000000000000000000000000003e3c33bc605e5d"); // 3,976,284
```
Comment (patricklodder) [inline src/chainparams.cpp:267]: I missed a `0x` prefix here. Not blocking but sloppy, so I will update this line.

Comment (patricklodder) [reply src/chainparams.cpp]: Done: [change after force push]( https://github.com/dogecoin/dogecoin/compare/0d3c023354e4a4e532fafc3da7c7ae472e221802..43c0008fb79dffa5971264c04aecc06c4576fa33)

**Comment by chromatic** [comment]
I've checked the numbers for mainnet and everything looks good to me. No long chaintips, confirmation numbers are sufficient, values match up appropriately.

**Review by chromatic — APPROVED** [review]
I verified all of the numbers via various means, including the estimates. All tests pass on 64-bit x86 Linux. ACK.
